File: app/src/main/java/com/jiangKlijna/IO.java
package com.jiangKlijna;

import java.io.*;

/**
 * io流工具类
 * Author: com.jiangKlijna
 */
public class IO {
    private IO() {
    }

    public static void copyFile_char(final File fsrc, final File descf) throws IOException {
        io(new FileReader(fsrc), new FileWriter(descf));
    }

    public static void copyFile_byte(final File fsrc, final File descf) throws IOException {
        io(new FileInputStream(fsrc), new FileOutputStream(descf));
    }

    public static void copyFile_char(final String src, final File descf) throws IOException {
        io(src, new FileWriter(descf));
    }

    public static void copyFile_byte(final String src, final File descf) throws IOException {
        io(src, new FileOutputStream(descf));
    }

    public static void copyFile_char(final byte[] bytes, final File descf) throws IOException {
        io(bytes, new FileWriter(descf));
    }

    public static void copyFile_byte(final byte[] bytes, final File descf) throws IOException {
        io(bytes, new FileOutputStream(descf));
    }

    public static String io(final Reader reader) throws IOException {
        StringWriter stringWriter = new StringWriter();
        io(reader, stringWriter);
        return stringWriter.toString();
    }

    public static String io(final InputStream reader) throws IOException {
        StringWriter stringWriter = new StringWriter();
        io(reader, stringWriter);
        return stringWriter.toString();
    }

    public static void io(final byte[] bytes, final Writer writer) throws IOException {
        io(new ByteArrayInputStream(bytes), writer);
    }

    public static void io(final byte[] bytes, final OutputStream writer) throws IOException {
        io(new ByteArrayInputStream(bytes), writer);
    }

    public static void io(final String string, final Writer writer) throws IOException {
        io(new StringReader(string), writer);
    }

    public static void io(final String string, final OutputStream writer) throws IOException {
        io(new StringReader(string), writer);
    }

    //字节流
    public static void io(final InputStream is, final OutputStream os) throws IOException {
        BufferedInputStream bis = new BufferedInputStream(is);
        BufferedOutputStream bos = new BufferedOutputStream(os);
        byte[] buf = new byte[2048];
        int len;
        while ((len = bis.read(buf)) != -1) {
            bos.write(buf, 0, len);
            bos.flush();
        }
        bis.close();
        bos.close();
    }

    public static void io(final InputStream is, final Writer os) throws IOException {
        io(new InputStreamReader(is), os);
    }

    public static void io(final Reader is, final OutputStream os) throws IOException {
        io(is, new OutputStreamWriter(os));
    }

    public static void io(final Reader is, final Writer os) throws IOException {
        BufferedReader bufr = new BufferedReader(is);
        BufferedWriter bufw = new BufferedWriter(os);
        String line;
        while ((line = bufr.readLine()) != null) {
            bufw.write(line);
            bufw.newLine();
            bufw.flush();
        }
        bufr.close();
        bufw.close();
    }

}

File: app/src/main/java/com/jiangKlijna/kotlin/IO.kt
package com.jiangKlijna.kotlin

import java.io.*

/**
 * io流工具类
 * Author: com.jiangKlijna
 */
object IO {

    private const val BYTE_ARRAY_LENGTH = 2048

    //文件相关
    @Throws(IOException::class)
    fun filec(fsrc: File, descf: File) {
        chars(FileReader(fsrc), FileWriter(descf))
    }

    @Throws(IOException::class)
    fun fileb(fsrc: File, descf: File) {
        bytes(FileInputStream(fsrc), FileOutputStream(descf))
    }

    @Throws(IOException::class)
    fun filec(src: String, descf: File) {
        str(src, FileWriter(descf))
    }

    @Throws(IOException::class)
    fun fileb(src: String, descf: File) {
        str(src, FileOutputStream(descf))
    }

    @Throws(IOException::class)
    fun filec(bytes: ByteArray, descf: File) {
        byt(bytes, FileWriter(descf))
    }

    @Throws(IOException::class)
    fun fileb(bytes: ByteArray, descf: File) {
        byt(bytes, FileOutputStream(descf))
    }

    @Throws(IOException::class)
    fun filec(descf: File): String {
        return str(FileReader(descf))
    }

    @Throws(IOException::class)
    fun fileb(descf: File): ByteArray {
        return byt(FileInputStream(descf))
    }

    //序列化对象
    @Throws(IOException::class)
    fun obj(obj: Any): ByteArray {
        val bo = ByteArrayOutputStream()
        val oo = ObjectOutputStream(bo)
        oo.writeObject(obj)
        val bytes = bo.toByteArray()
        bo.close()
        oo.close()
        return bytes
    }

    @Throws(Exception::class)
    fun <T> obj(bytes: ByteArray): T {
        val sin = ObjectInputStream(ByteArrayInputStream(bytes))
        return sin.readObject() as T
    }
    
    //字符串相关
    @Throws(IOException::class)
    fun str(string: String, writer: Writer) {
        chars(StringReader(string), writer)
    }

    @Throws(IOException::class)
    fun str(string: String, writer: OutputStream) {
        chars_bytes(StringReader(string), writer)
    }

    @Throws(IOException::class)
    fun str(reader: Reader): String {
        val stringWriter = StringWriter()
        chars(reader, stringWriter)
        return stringWriter.toString()
    }

    @Throws(IOException::class)
    fun str(reader: InputStream): String {
        val stringWriter = StringWriter()
        bytes_chars(reader, stringWriter)
        return stringWriter.toString()
    }

    //字节数组相关
    @Throws(IOException::class)
    fun byt(bytes: ByteArray, writer: Writer) {
        bytes_chars(ByteArrayInputStream(bytes), writer)
    }

    @Throws(IOException::class)
    fun byt(bytes: ByteArray, writer: OutputStream) {
        bytes(ByteArrayInputStream(bytes), writer)
    }

    @Throws(IOException::class)
    fun byt(reader: InputStream): ByteArray {
        val byteArrayOutputStream = ByteArrayOutputStream()
        bytes(reader, byteArrayOutputStream)
        return byteArrayOutputStream.toByteArray()
    }

    @Throws(IOException::class)
    fun byt(reader: Reader): ByteArray {
        val byteArrayOutputStream = ByteArrayOutputStream()
        chars_bytes(reader, byteArrayOutputStream)
        return byteArrayOutputStream.toByteArray()
    }

    //字节流<->字符流
    @Throws(IOException::class)
    fun bytes_chars(`is`: InputStream, os: Writer) {
        chars(InputStreamReader(`is`), os)
    }

    @Throws(IOException::class)
    fun chars_bytes(`is`: Reader, os: OutputStream) {
        chars(`is`, OutputStreamWriter(os))
    }

    //字节流
    @Throws(IOException::class)
    fun bytes(`is`: InputStream, os: OutputStream) {
        val bis = BufferedInputStream(`is`)
        val bos = BufferedOutputStream(os)
        val buf = ByteArray(BYTE_ARRAY_LENGTH)
        var len: Int
        while (true) {
            len = bis.read()
            if (len == -1) break
            else {
                bos.write(buf, 0, len)
                bos.flush()
            }
        }
        bis.close()
        bos.close()
    }

    //字符流
    @Throws(IOException::class)
    fun chars(`is`: Reader, os: Writer) {
        val bufr = BufferedReader(`is`)
        val bufw = BufferedWriter(os)
        var line: String
        while (true) {
            line = bufr.readLine() ?: break
            bufw.write(line)
            bufw.newLine()
            bufw.flush()
        }
        bufr.close()
        bufw.close()
    }

}
